File: hive/core/rules.py
"""Movement rules for each Hive piece type."""

from __future__ import annotations
from .hex import Hex, DIRECTIONS
from .board import Board
from .pieces import Piece, PieceType, PieceColor


def get_moves(piece: Piece, board: Board) -> set[Hex]:
    """Get all legal destination hexes for a piece already on the board."""
    pos = board.piece_position(piece)
    if pos is None:
        return set()

    # Piece must be on top of its stack to move
    if board.top_piece(pos) != piece:
        return set()

    # One Hive Rule: removing this piece must not disconnect the hive
    # (Beetles on top of stacks don't break connectivity when they move)
    if board.stack_height(pos) == 1:
        if not board.is_connected(exclude=piece):
            return set()

    move_fn = _MOVE_FUNCTIONS.get(piece.piece_type)
    if move_fn is None:
        return set()

    return move_fn(piece, pos, board)


def get_placements(color: PieceColor, board: Board) -> set[Hex]:
    """Get all legal placement positions for a player.

    A new piece must be placed adjacent to at least one friendly piece
    and NOT adjacent to any enemy piece. Exception: on the second move
    of the game (total move 2), the constraint about enemy adjacency
    is relaxed.
    """
    my_pieces = board.pieces_on_board(color)
    opp_color = PieceColor.BLACK if color == PieceColor.WHITE else PieceColor.WHITE
    opp_pieces = board.pieces_on_board(opp_color)

    # First piece: place at origin
    if not my_pieces and not opp_pieces:
        return {Hex(0, 0)}

    # Second piece (first move by opponent): adjacent to the first piece
    if not my_pieces and opp_pieces:
        result = set()
        for p in opp_pieces:
            pos = board.piece_position(p)
            result.update(board.empty_neighbors(pos))
        return result

    # Normal placement: adjacent to friendly, not adjacent to enemy
    friendly_positions = {board.piece_position(p) for p in my_pieces}
    enemy_positions = {board.piece_position(p) for p in opp_pieces}

    # All empty hexes adjacent to friendly pieces
    candidates = set()
    for fpos in friendly_positions:
        candidates.update(board.empty_neighbors(fpos))

    # Filter out those adjacent to enemy pieces
    valid = set()
    for cand in candidates:
        adjacent_to_enemy = False
        for n in cand.neighbors():
            top = board.top_piece(n)
            if top is not None and top.color == opp_color:
                adjacent_to_enemy = True
                break
        if not adjacent_to_enemy:
            valid.add(cand)

    return valid


# ---- Piece-specific movement ----

def _queen_moves(piece: Piece, pos: Hex, board: Board) -> set[Hex]:
    """Queen: moves exactly 1 space by sliding."""
    board.remove_piece(piece)
    moves = set()
    for n in pos.neighbors():
        if n not in board.occupied and board.can_slide(pos, n):
            # Must remain adjacent to hive after moving
            if any(adj in board.occupied for adj in n.neighbors()):
                moves.add(n)
    board.place_piece(piece, pos)
    return moves


def _beetle_moves(piece: Piece, pos: Hex, board: Board) -> set[Hex]:
    """Beetle: moves 1 space, can climb on top of the hive."""
    board.remove_piece(piece)
    moves = set()
    for n in pos.neighbors():
        # Beetle can move to occupied or empty adjacent hexes
        src_height = board.stack_height(pos)  # height after removing beetle
        dst_height = board.stack_height(n)

        # If both source and dest are ground level, normal slide rules apply
        if src_height == 0 and dst_height == 0:
            if board.can_slide(pos, n):
                # Must stay adjacent to hive
                if any(adj in board.occupied for adj in n.neighbors()):
                    moves.add(n)
        else:
            # Moving up, down, or across the top of the hive
            # Gate rule for elevated movement: need to check if the beetle
            # can physically pass between the two common neighbors
            # At height, the beetle can always move (no gate blocking at elevation)
            # Actually: gate blocking applies based on max height of common neighbors
            # vs max(src_height, dst_height). Simplified: if beetle or target is elevated,
            # gate doesn't block.
            if dst_height > 0 or any(adj in board.occupied for adj in n.neighbors()):
                moves.add(n)

    board.place_piece(piece, pos)
    return moves


def _grasshopper_moves(piece: Piece, pos: Hex, board: Board) -> set[Hex]:
    """Grasshopper: jumps in a straight line over at least one piece."""
    board.remove_piece(piece)
    moves = set()
    for d in DIRECTIONS:
        # Must jump over at least one piece
        current = pos + d
        if current not in board.occupied:
            continue
        # Keep going until we find an empty hex
        while current in board.occupied:
            current = current + d
        moves.add(current)
    board.place_piece(piece, pos)
    return moves


def _spider_moves(piece: Piece, pos: Hex, board: Board) -> set[Hex]:
    """Spider: moves exactly 3 spaces by sliding along the hive."""
    board.remove_piece(piece)
    moves = _walk_exactly_n(pos, 3, board)
    board.place_piece(piece, pos)
    return moves


def _ant_moves(piece: Piece, pos: Hex, board: Board) -> set[Hex]:
    """Ant: slides any number of spaces around the hive perimeter."""
    board.remove_piece(piece)
    moves = _walk_any(pos, board)
    board.place_piece(piece, pos)
    return moves


def _walk_exactly_n(start: Hex, n: int, board: Board) -> set[Hex]:
    """Find all hexes reachable by sliding exactly n steps along the hive edge."""
    # BFS tracking (position, steps_taken), visited per step to allow revisiting at different steps
    results = set()
    # State: (position, steps_taken), track visited as set of (pos, step) to prevent loops
    frontier = [(start, 0, frozenset([start]))]

    while frontier:
        pos, steps, visited = frontier.pop()
        if steps == n:
            if pos != start:
                results.add(pos)
            continue

        for neighbor in pos.neighbors():
            if neighbor in board.occupied:
                continue
            if neighbor in visited:
                continue
            # Must be able to slide there
            if not board.can_slide(pos, neighbor):
                continue
            # Must remain adjacent to the hive
            if not any(adj in board.occupied for adj in neighbor.neighbors()):
                continue
            frontier.append((neighbor, steps + 1, visited | {neighbor}))

    return results


def _walk_any(start: Hex, board: Board) -> set[Hex]:
    """Find all hexes reachable by sliding any number of steps (ant movement)."""
    visited = {start}
    queue = [start]
    results = set()

    while queue:
        pos = queue.pop()
        for neighbor in pos.neighbors():
            if neighbor in board.occupied:
                continue
            if neighbor in visited:
                continue
            if not board.can_slide(pos, neighbor):
                continue
            if not any(adj in board.occupied for adj in neighbor.neighbors()):
                continue
            visited.add(neighbor)
            results.add(neighbor)
            queue.append(neighbor)

    return results


_MOVE_FUNCTIONS = {
    PieceType.QUEEN: _queen_moves,
    PieceType.BEETLE: _beetle_moves,
    PieceType.GRASSHOPPER: _grasshopper_moves,
    PieceType.SPIDER: _spider_moves,
    PieceType.ANT: _ant_moves,
}

File: rust/crates/hive-game/src/rules.rs
/// Movement rules for each Hive piece type.

use crate::hex::{Hex, DIRECTIONS, hex_add, hex_neighbors};
use crate::board::Board;
use crate::piece::{Piece, PieceColor, PieceType};

/// Get all legal destination hexes for a piece already on the board.
/// Temporarily removes the piece from the board to compute moves, then restores it.
pub fn get_moves(piece: Piece, board: &mut Board, articulation_points: &[Hex]) -> Vec<Hex> {
    let pos = match board.piece_position(piece) {
        Some(p) => p,
        None => return Vec::new(),
    };

    // Piece must be on top of its stack
    if board.top_piece(pos) != Some(piece) {
        return Vec::new();
    }

    // One Hive Rule: pieces on stacks can always move
    if board.stack_height(pos) == 1 {
        if articulation_points.contains(&pos) {
            return Vec::new();
        }
    }

    // Temporarily remove piece from board
    board.remove_piece(piece).unwrap();

    let moves = match piece.piece_type() {
        PieceType::Queen => queen_moves(pos, board),
        PieceType::Spider => spider_moves(pos, board),
        PieceType::Beetle => beetle_moves(pos, board),
        PieceType::Grasshopper => grasshopper_moves(pos, board),
        PieceType::Ant => ant_moves(pos, board),
    };

    // Restore piece
    board.place_piece(piece, pos).unwrap();

    moves
}

/// Get all legal placement positions for a player.
pub fn get_placements(color: PieceColor, board: &Board) -> Vec<Hex> {
    let my_pieces = board.pieces_on_board(color);
    let opp_color = color.opposite();
    let opp_pieces = board.pieces_on_board(opp_color);

    // First piece: place at origin
    if my_pieces.is_empty() && opp_pieces.is_empty() {
        return vec![(0, 0)];
    }

    // Second piece: adjacent to existing pieces
    if my_pieces.is_empty() && !opp_pieces.is_empty() {
        let mut result = Vec::new();
        for p in &opp_pieces {
            let pos = board.piece_position(*p).unwrap();
            for n in board.empty_neighbors(pos) {
                if !result.contains(&n) {
                    result.push(n);
                }
            }
        }
        return result;
    }

    // Normal: adjacent to friendly, not adjacent to enemy
    let friendly_positions: Vec<Hex> = my_pieces
        .iter()
        .filter_map(|p| board.piece_position(*p))
        .collect();

    // Collect all empty hexes adjacent to friendly pieces
    let mut candidates = Vec::new();
    for &fpos in &friendly_positions {
        for n in board.empty_neighbors(fpos) {
            if !candidates.contains(&n) {
                candidates.push(n);
            }
        }
    }

    // Filter out those adjacent to enemy pieces
    let mut valid = Vec::new();
    for cand in candidates {
        let mut adjacent_to_enemy = false;
        for &n in hex_neighbors(cand).iter() {
            if let Some(top) = board.top_piece(n) {
                if top.color() == opp_color {
                    adjacent_to_enemy = true;
                    break;
                }
            }
        }
        if !adjacent_to_enemy {
            valid.push(cand);
        }
    }

    valid
}

// ---- Piece-specific movement ----
// All functions below assume the piece has already been removed from `pos`.

fn queen_moves(pos: Hex, board: &Board) -> Vec<Hex> {
    let mut moves = Vec::new();
    for &n in hex_neighbors(pos).iter() {
        if !board.is_occupied(n) && board.can_slide(pos, n) {
            if hex_neighbors(n).iter().any(|&adj| board.is_occupied(adj)) {
                moves.push(n);
            }
        }
    }
    moves
}

fn beetle_moves(pos: Hex, board: &Board) -> Vec<Hex> {
    let mut moves = Vec::new();
    let src_height = board.stack_height(pos);

    for &n in hex_neighbors(pos).iter() {
        let dst_height = board.stack_height(n);

        if src_height == 0 && dst_height == 0 {
            // Ground level: normal slide rules
            if board.can_slide(pos, n) {
                if hex_neighbors(n).iter().any(|&adj| board.is_occupied(adj)) {
                    moves.push(n);
                }
            }
        } else {
            // Elevated movement: no gate blocking
            if dst_height > 0 || hex_neighbors(n).iter().any(|&adj| board.is_occupied(adj)) {
                moves.push(n);
            }
        }
    }

    moves
}

fn grasshopper_moves(pos: Hex, board: &Board) -> Vec<Hex> {
    let mut moves = Vec::new();
    for &d in DIRECTIONS.iter() {
        let mut current = hex_add(pos, d);
        if !board.is_occupied(current) {
            continue; // must jump over at least one piece
        }
        while board.is_occupied(current) {
            current = hex_add(current, d);
        }
        moves.push(current);
    }
    moves
}

fn spider_moves(pos: Hex, board: &Board) -> Vec<Hex> {
    walk_exactly_n(pos, 3, board)
}

fn ant_moves(pos: Hex, board: &Board) -> Vec<Hex> {
    walk_any(pos, board)
}

/// Find all hexes reachable by sliding exactly n steps.
fn walk_exactly_n(start: Hex, n: usize, board: &Board) -> Vec<Hex> {
    let mut results = Vec::new();

    struct State {
        pos: Hex,
        steps: usize,
        visited: Vec<Hex>,
    }

    let mut frontier = vec![State {
        pos: start,
        steps: 0,
        visited: vec![start],
    }];

    while let Some(state) = frontier.pop() {
        if state.steps == n {
            if state.pos != start && !results.contains(&state.pos) {
                results.push(state.pos);
            }
            continue;
        }

        for &neighbor in hex_neighbors(state.pos).iter() {
            if board.is_occupied(neighbor) {
                continue;
            }
            if state.visited.contains(&neighbor) {
                continue;
            }
            if !board.can_slide(state.pos, neighbor) {
                continue;
            }
            if !hex_neighbors(neighbor).iter().any(|&adj| board.is_occupied(adj)) {
                continue;
            }
            let mut new_visited = state.visited.clone();
            new_visited.push(neighbor);
            frontier.push(State {
                pos: neighbor,
                steps: state.steps + 1,
                visited: new_visited,
            });
        }
    }

    results
}

/// Find all hexes reachable by sliding any number of steps (ant movement).
fn walk_any(start: Hex, board: &Board) -> Vec<Hex> {
    let mut visited = vec![start];
    let mut results = Vec::new();
    let mut queue = vec![start];

    while let Some(pos) = queue.pop() {
        for &neighbor in hex_neighbors(pos).iter() {
            if board.is_occupied(neighbor) {
                continue;
            }
            if visited.contains(&neighbor) {
                continue;
            }
            if !board.can_slide(pos, neighbor) {
                continue;
            }
            if !hex_neighbors(neighbor).iter().any(|&adj| board.is_occupied(adj)) {
                continue;
            }
            visited.push(neighbor);
            results.push(neighbor);
            queue.push(neighbor);
        }
    }

    results
}

#[cfg(test)]
#[allow(unused_must_use)]
mod tests {
    use super::*;

    fn setup_basic_board() -> Board {
        let mut board = Board::new();
        // Place white queen at origin and black queen adjacent
        let wq = Piece::new(PieceColor::White, PieceType::Queen, 1);
        let bq = Piece::new(PieceColor::Black, PieceType::Queen, 1);
        board.place_piece(wq, (0, 0));
        board.place_piece(bq, (1, 0));
        board
    }

    #[test]
    fn test_first_placement() {
        let board = Board::new();
        let placements = get_placements(PieceColor::White, &board);
        assert_eq!(placements, vec![(0, 0)]);
    }

    #[test]
    fn test_second_placement() {
        let mut board = Board::new();
        let wq = Piece::new(PieceColor::White, PieceType::Queen, 1);
        board.place_piece(wq, (0, 0));

        let placements = get_placements(PieceColor::Black, &board);
        assert_eq!(placements.len(), 6);
    }

    #[test]
    fn test_queen_moves() {
        let mut board = setup_basic_board();
        let wq = Piece::new(PieceColor::White, PieceType::Queen, 1);
        let moves = get_moves(wq, &mut board, &[]);
        // Queen should be able to slide to 2 positions adjacent to both pieces
        assert!(!moves.is_empty());
    }

    #[test]
    fn test_grasshopper_jump() {
        let mut board = Board::new();
        let wq = Piece::new(PieceColor::White, PieceType::Queen, 1);
        let bq = Piece::new(PieceColor::Black, PieceType::Queen, 1);
        let wg = Piece::new(PieceColor::White, PieceType::Grasshopper, 1);

        board.place_piece(wq, (0, 0));
        board.place_piece(bq, (1, 0));
        board.place_piece(wg, (-1, 0));

        let moves = get_moves(wg, &mut board, &[]);
        // Should jump over wq and bq to land at (2, 0)
        assert!(moves.contains(&(2, 0)));
    }

    #[test]
    fn test_beetle_stacking() {
        let mut board = Board::new();
        let wq = Piece::new(PieceColor::White, PieceType::Queen, 1);
        let bq = Piece::new(PieceColor::Black, PieceType::Queen, 1);
        let wb = Piece::new(PieceColor::White, PieceType::Beetle, 1);

        board.place_piece(wq, (0, 0));
        board.place_piece(bq, (1, 0));
        board.place_piece(wb, (0, -1)); // adjacent to wq

        // Add more pieces so beetle isn't an articulation point
        let ws = Piece::new(PieceColor::White, PieceType::Spider, 1);
        board.place_piece(ws, (1, -1));

        let moves = get_moves(wb, &mut board, &[]);
        // Beetle should be able to move onto occupied hexes
        assert!(moves.contains(&(0, 0)) || moves.contains(&(1, 0)));
    }

    #[test]
    fn test_articulation_point_blocks_move() {
        let mut board = Board::new();
        let wq = Piece::new(PieceColor::White, PieceType::Queen, 1);
        let bq = Piece::new(PieceColor::Black, PieceType::Queen, 1);
        let wa = Piece::new(PieceColor::White, PieceType::Ant, 1);

        board.place_piece(wq, (0, 0));
        board.place_piece(bq, (1, 0));
        board.place_piece(wa, (2, 0));

        let aps = board.articulation_points();
        // bq at (1,0) is an articulation point
        let moves = get_moves(bq, &mut board, &aps);
        assert!(moves.is_empty());
    }
}
